javascript/extractor/src/com/semmle/js/extractor/CFGExtractor.java

```diff
@@ -210,8 +242,21 @@ public CFGExtractor(ASTExtractor astExtractor) {
       return xs;
     }
     
-    Set<Node> set = new HashSet<>();
-    return Stream.concat(xs.stream(), ys.stream()).filter(set::add).collect(Collectors.toList());
+    return nonNullUnion(xs, ys);
+  }
+
+  /**
+   * Creates an order preserving concatenation of the nodes in `xs` and `ys` without duplicates.
+   * Where `xs` and `ys` have non null values, and are non-empty.
+   */
+  private static Collection<Node> nonNullUnion(Collection<Node> xs, Collection<Node> ys) {
+    List<Node> result = new ArrayList<>(xs);
+    for (Node y : ys) {
+      if (!result.contains(y)) {
```

Comment:
This is `O(n^2)`.

Reply:
I know, but `n` is always small. 

I did an experiment trying to extract `Microsoft/vscode`, and `n` was at most 3.  

My idea was to avoid allocating a `HashSet`.  
Also, it's basically what the implementation did before my refactor.